src/core/tests/unittests/pickingcontroller-test.cpp

```diff
@@ -564,4 +965,106 @@ TEST(PickingControllerTest, MouseDrag0to1) {
     }
 }
 
+TEST(PickingControllerTest, TouchDrag0to1) {
+    PickingManager pm;
+    pm.registerPickingAction(
+        nullptr, [](PickingEvent*) {}, 2);
+    pm.registerPickingAction(
+        nullptr, [](PickingEvent*) {}, 3);
+    PickingControllerTouchState ms(&pm);
+    TouchDevice device{TouchDevice::DeviceType::TouchPad};
+    {
+        uvec2 pos{1, 1};  // id = 0
+        auto event = touchEvent(TouchState::Started, pos, dim, device);
+
+        TestPropagator tp;
+        ms.propagateEvent(&event, &tp, canvas[dim.y - pos.y][pos.x]);
+        ASSERT_EQ(tp.events.size(), 0);
+    }
+    {
+        SCOPED_TRACE("Press 1,1");
+        // uvec2 pre{1, 1};  // id = 0
+        uvec2 pos{1, 1};  // id = 0
+
+        auto event = touchEvent(TouchState::Stationary, pos, dim, device);
+        TestPropagator tp;
+        ms.propagateEvent(&event, &tp, canvas[dim.y - pos.y][pos.x]);
+        ASSERT_EQ(tp.events.size(), 0);
+    }
+
+    {
+        SCOPED_TRACE("Drag 1,1 -> 3,3");
+        // uvec2 pre{1, 1};  // id = 0
+        uvec2 pos{3, 3};  // id = 1
+
+        auto event = touchEvent(TouchState::Stationary, pos, dim, device);
+        TestPropagator tp;
+        ms.propagateEvent(&event, &tp, canvas[dim.y - pos.y][pos.x]);
+        ASSERT_EQ(tp.events.size(), 0);
+    }
+    {
+        SCOPED_TRACE("Release 3,3");
+        // uvec2 pre{3, 3};  // id = 1
+        uvec2 pos{3, 3};  // id = 1
+
+        auto event = touchEvent(TouchState::Updated, pos, dim, device);
+        TestPropagator tp;
+        ms.propagateEvent(&event, &tp, canvas[dim.y - pos.y][pos.x]);
+        ASSERT_EQ(tp.events.size(), 1);
+        // NOTE: PPI is changed to None, should it really be Primary as is the case for Mouse
+        testPickingEvent(tp.events[0].get(), PS::Started, PHS::Enter, PPS::None, PPI::None,
+                         PPIs{PickingPressItem::Primary}, 2, 2, 0, 0, 1, 1, 1, 1, ndc(pos),
+                         dvec3{0.0}, dvec3{0.0});
+    }
+}
+
+TEST(PickingControllerTest, TouchScreenDrag0to1) {
+    PickingManager pm;
+    pm.registerPickingAction(
+        nullptr, [](PickingEvent*) {}, 2);
+    pm.registerPickingAction(
+        nullptr, [](PickingEvent*) {}, 3);
+    PickingControllerTouchState ms(&pm);
+    TouchDevice device{TouchDevice::DeviceType::TouchScreen};
+    {
+        uvec2 pos{1, 1};  // id = 0
+        auto event = touchEvent(TouchState::Started, pos, dim, device);
+
+        TestPropagator tp;
+        ms.propagateEvent(&event, &tp, canvas[dim.y - pos.y][pos.x]);
+        ASSERT_EQ(tp.events.size(), 0);
+    }
+    //{
+    //    SCOPED_TRACE("Press 1,1");
+    //    // uvec2 pre{1, 1};  // id = 0
+    //    uvec2 pos{1, 1};  // id = 0
+
+    //    auto event = touchEvent(TouchState::Stationary, pos, dim, device);
+    //    TestPropagator tp;
+    //    ms.propagateEvent(&event, &tp, canvas[dim.y - pos.y][pos.x]);
+    //    ASSERT_EQ(tp.events.size(), 0);
+    //}
```

Comment:
?

Reply:
First touch on TouchScreen will always press. Removed the commented part